Client-side translations: gettext and plural support

Ekdohibs opened the pull request:
This is a rebase of #12446.

This PR is complete, see instructions on the previous PR for how to test.

## How to test

For now, take a mod using client-side translations, and replace the .tr file with a .po one, with the following entries for each original translated string:
```
msgctx "<textdomain>"
msgid "<original string>"
msgstr "<translated string>"
```

rubenwardy commented:
I will have time to review this soon. I've given it a quick look through, and have too comments:

* The parser is in the Translation class, would be nice to have a separate class for this with unit tests. Although, I suppose same for .tr
* I wonder if the script is outdated now that we have a util script for extracting .tr

Ekdohibs commented:
I just rebased it, it should be good now. I'll try to write some unit tests and move the parser to its own class, it might be a bit harder to do for `.mo` files as they are a binary format.
Concerning the script, I can't seem to find the util script in the minetest/ repository, has it been merged there? The objective of including it was so that it is distributed with mintest itself; it is also updated to support plural translations and `.po` output. If the script is hosted somewhere else, I can make the PR there, however.

y5nw commented on src/translation.cpp:
```diff
@@ -145,11 +211,337 @@ void Translations::loadTranslation(const std::string &data)
+			line.resize(line.length() - 1);
+
+		if (line.empty() || line[0] == '#')
+			continue;
```
Since `xgettext` can generate [fuzzy entries](https://www.gnu.org/software/gettext/manual/html_node/Fuzzy-Entries.html) it would make sense to check for (and skip) those.

Ekdohibs replied:
From the documentation of gettext, it seems to say that fuzzy entries are considered as normal entries for most purposes, so I wouldn't drop them. In particular, the `.po` to `.mo` compilation would use them, so we should use them as well if we don't want to introduce a discrepancy.

y5nw replied:
> From the documentation of gettext, it seems to say that fuzzy entries are considered as normal entries for most purposes, so I wouldn't drop them.

While that is the case, the problem with fuzzy matching is that it is enabled by default in `msgmerge` (unless the `-N` flag is passed) and can often match entries with different (sometimes opposite, e.g. "enabled" and "disabled") meanings while assigning the same translation if one is missing, and there are also reasons to generate fuzzy entries, e.g. if the source string does not change much (adding a period to the end of the source string) or to give translators a base string to modify/work with.

> In particular, the `.po` to `.mo` compilation would use them, so we should use them as well if we don't want to introduce a discrepancy.

`msgfmt` does not seem to include fuzzy entries by default (there is the [`-f` flag](https://www.gnu.org/software/gettext/manual/html_node/msgfmt-Invocation.html#Input-file-interpretation-1) option for including them which GNU does not recommend using).

Ekdohibs replied:
Thanks for the documentation! I'll implement matching the fuzzy entries in that case.

y5nw commented on doc/lua_api.md:
```diff
@@ -4119,6 +4137,14 @@ The file should be a text file, with the following format:
+Gettext files can also be used as translations. A translation file has the suffix
+`.[lang].po` or `.[lang].mo`, depending on whether it is compiled or not, and must
+also be placed in the `locale` subdirectory of the mod. The value of `textdomain`
+is `msgctx` in the gettext files.
```
Wouldn't it make sense to also allow placing `.po` files in the `po/` subdirectory of the mod?

rubenwardy replied:
```suggestion
is `msgctxt` in the gettext files.
```

rubenwardy replied:
Please add an example:

```po
msgctxt "textdomain"

msgid "Hello world!"
msgstr "Bonjour le monde!"
```

y5nw reviewed (commented):
Nitpicking.

rubenwardy commented:
> Concerning the script, I can't seem to find the util script in the minetest/ repository, has it been merged there? The objective of including it was so that it is distributed with mintest itself; it is also updated to support plural translations and .po output. If the script is hosted somewhere else, I can make the PR there, however.

I was referring to https://github.com/minetest/minetest/blob/master/util/mod_translation_updater.py

Makes sense for someone to update that in another PR though

rubenwardy commented on games/devtest/mods/testtranslations/init.lua:
```diff
@@ -0,0 +1,17 @@
+	description = "Test translations",
+	privs = {},
+	func = function(name, param)
+		minetest.chat_send_player(name, "Please ensure your locale is set to \"fr\"")
```
```suggestion
		if minetest.get_player_information(name).language ~= "fr" then
				return false, "Please restart Minetest with language = fr"
		end

```

Ekdohibs replied:
I applied the suggestion, but I must have forgotten to pull before rebasing, deleting the commit. I'll commit it again, thanks for catching it!
Although, now that I'm thinking about it, we really want to test this both with `fr` and default locale (to see both untranslated and translated strings), I'll need to check what `lang_code` is in that case.

y5nw commented on src/translation.cpp:
```diff
@@ -145,11 +211,337 @@ void Translations::loadTranslation(const std::string &data)
+	if (ctx != entry.end()) {
+		textdomain = ctx->second;
+	} else {
+		textdomain = L"";
```
Since other PO file generators do not necessarily generate `msgctxt` entries it would IMO be a good idea to use the base name of the file (i.e. without the `LANG.po` suffix) as the textdomain if a `msgctxt` entry is not available.

Ekdohibs replied:
Done

Ekdohibs commented on src/translation.h:
```diff
@@ -30,11 +32,26 @@ extern Translations *g_client_translations;
 
 private:
 	std::unordered_map<std::wstring, std::wstring> m_translations;
+	std::unordered_map<std::wstring, std::vector<std::wstring>> m_plural_translations;
```
Plural translations are rare and have an extra indirection due to the vector; I feel like the added complexity is sufficiently small to be worth the cost.

Ekdohibs commented on doc/lua_api.md:
````diff
@@ -4078,19 +4080,35 @@ Two functions are provided to translate strings: `minetest.translate` and
 
   ```lua
-  local S = minetest.get_translator()
+  local S, NS = minetest.get_translator()
````
Done.

Ekdohibs commented:
Hmmmm, I'm not why the CI fails, do you have an idea?

rubenwardy commented:
Here's a patch to fix content translation with .po files:

```
commit 484cb1bab5c080f4cc1fdd3e409731c62f19b277
Author: rubenwardy <rw@rubenwardy.com>
Date:   Sun Mar 24 18:24:30 2024 +0000

    Fix content translation and Gettext format

diff --git a/src/gui/guiEngine.cpp b/src/gui/guiEngine.cpp
index 4e22c3ae8..678f8a64f 100644
--- a/src/gui/guiEngine.cpp
+++ b/src/gui/guiEngine.cpp
@@ -214,15 +214,29 @@ GUIEngine::GUIEngine(JoystickController *joystick,
 
 
 /******************************************************************************/
-std::string findLocaleFileInMods(const std::string &path, const std::string &filename)
+std::string findLocaleFileWithExtension(const std::string &path)
+{
+	errorstream << "Finding " << path << std::endl;
+	if (fs::PathExists(path + ".mo"))
+		return path + ".mo";
+	if (fs::PathExists(path + ".po"))
+		return path + ".po";
+	if (fs::PathExists(path + ".tr"))
+		return path + ".tr";
+	return "";
+}
+
+
+/******************************************************************************/
+std::string findLocaleFileInMods(const std::string &path, const std::string &filename_no_ext)
 {
 	std::vector<ModSpec> mods = flattenMods(getModsInPath(path, "root", true));
 
 	for (const auto &mod : mods) {
-		std::string ret = mod.path + DIR_DELIM "locale" DIR_DELIM + filename;
-		if (fs::PathExists(ret)) {
+		std::string ret = findLocaleFileWithExtension(
+				mod.path + DIR_DELIM "locale" DIR_DELIM + filename_no_ext);
+		if (!ret.empty())
 			return ret;
-		}
 	}
 
 	return "";
@@ -235,8 +249,8 @@ Translations *GUIEngine::getContentTranslations(const std::string &path,
 	if (domain.empty() || lang_code.empty())
 		return nullptr;
 
-	std::string filename = domain + "." + lang_code + ".tr";
-	std::string key = path + DIR_DELIM "locale" DIR_DELIM + filename;
+	std::string filename_no_ext = domain + "." + lang_code;
+	std::string key = path + DIR_DELIM "locale" DIR_DELIM + filename_no_ext;
 
 	if (key == m_last_translations_key)
 		return &m_last_translations;
@@ -244,10 +258,11 @@ Translations *GUIEngine::getContentTranslations(const std::string &path,
 	std::string trans_path = key;
 	ContentType type = getContentType(path);
 	if (type == ContentType::GAME)
-		trans_path = findLocaleFileInMods(path + DIR_DELIM "mods" DIR_DELIM, filename);
+		trans_path = findLocaleFileInMods(path + DIR_DELIM "mods" DIR_DELIM, filename_no_ext);
 	else if (type == ContentType::MODPACK)
-		trans_path = findLocaleFileInMods(path, filename);
-	// We don't need to search for locale files in a mod, as there's only one `locale` folder.
+		trans_path = findLocaleFileInMods(path, filename_no_ext);
+	else
+		trans_path = findLocaleFileWithExtension(trans_path);
 
 	if (trans_path.empty())
 		return nullptr;
```

rubenwardy commented:
Here's a zip file for testing gettext content translation

[gettext_content_translation.zip](https://github.com/minetest/minetest/files/14736402/gettext_content_translation.zip)

rubenwardy commented:
> The parser is in the Translation class, would be nice to have a separate class for this with unit tests. Although, I suppose same for .tr

The reason I say this is because it's a good idea to have unit tests when writing a parser for a file format as it gives some confidence that the parser actually works

rubenwardy commented on games/devtest/mods/testtranslations/translation_mo.fr.po:
```diff
@@ -0,0 +1,9 @@
+msgid "Testing .mo files: untranslated"
+msgstr "Testing .mo files: translated"
+
+msgctxt "testtranslations"
```
If this msgctxt is required for every message, this should be checked by the engine and warned/errored about

y5nw replied:
I assume this file is used to generate `locale/translation_mo.fr.mo`? The translation file reader uses the base filename (here: `translation_mo`) as the text domain if one is not specified.

The [GNU Gettext doc on PO files](https://www.gnu.org/software/gettext/manual/html_node/PO-Files.html) does not seem clear about about the scope (i.e. the entries that the `msgctxt` applies to) of `msgctxt`. However, considering that
* The documentation states that "an empty context string and an absent msgctxt line do not mean the same thing";
* Entries with `msgctxt` are considered as "entries with a context specifier"; and
* The entire documentation does not mention "scopes" of other specifiers (that said, having `msgid` and `msgtxt` to multiple entries would make little sense anyway)

It would be safe to assume that entries without an explicit context specifier would be considered as entries without context information, which, combined with the fact that Minetest does not differentiate between "text domain" and "context" (and, in this PR, uses the former if the latter is absent), means that any entry without an explicit `msgctxt` specifier would be put under the default (here: `translation_mo`) domain, which would explain the duplicate `msgctxt`.

It is probably also worth noting that the generated `.mo` file includes context information on every entry as well.

On a less related note, while entries without `msgctxt` are allowed (see https://github.com/minetest/minetest/pull/13687#discussion_r1518407844), there does not seem to be any testcase/testfile related to this.
